File: codex-rs/core/src/tools/handlers/mcp.rs
use async_trait::async_trait;

use crate::function_tool::FunctionCallError;
use crate::tools::context::ToolInvocation;
use crate::tools::context::ToolOutput;
use crate::tools::context::ToolPayload;
use crate::tools::orchestrator::ToolOrchestrator;
use crate::tools::registry::ToolHandler;
use crate::tools::registry::ToolKind;
use crate::tools::runtimes::mcp::McpRuntime;
use crate::tools::runtimes::mcp::McpToolCallRequest;
use crate::tools::sandboxing::ToolCtx;
use crate::tools::sandboxing::ToolError;

pub struct McpHandler;

#[async_trait]
impl ToolHandler for McpHandler {
    fn kind(&self) -> ToolKind {
        ToolKind::Mcp
    }

    async fn handle(&self, invocation: ToolInvocation) -> Result<ToolOutput, FunctionCallError> {
        let ToolInvocation {
            session,
            turn,
            call_id,
            tool_name,
            payload,
            ..
        } = invocation;

        let payload = match payload {
            ToolPayload::Mcp {
                server,
                tool,
                raw_arguments,
            } => (server, tool, raw_arguments),
            _ => {
                return Err(FunctionCallError::RespondToModel(
                    "mcp handler received unsupported payload".to_string(),
                ));
            }
        };

        let (server, tool, raw_arguments) = payload;
        let req = McpToolCallRequest {
            server,
            tool,
            raw_arguments,
            cwd: turn.cwd.clone(),
        };

        let mut orchestrator = ToolOrchestrator::new();
        let mut runtime = McpRuntime::new();
        let tool_ctx = ToolCtx {
            session: session.as_ref(),
            turn: turn.as_ref(),
            call_id: call_id.clone(),
            tool_name: tool_name.to_string(),
        };

        let response = orchestrator
            .run(&mut runtime, &req, &tool_ctx, &turn, turn.approval_policy)
            .await
            .map_err(map_tool_error)?;

        match response {
            codex_protocol::models::ResponseInputItem::McpToolCallOutput { result, .. } => {
                Ok(ToolOutput::Mcp { result })
            }
            codex_protocol::models::ResponseInputItem::FunctionCallOutput { output, .. } => {
                let codex_protocol::models::FunctionCallOutputPayload {
                    content,
                    content_items,
                    success,
                } = output;
                Ok(ToolOutput::Function {
                    content,
                    content_items,
                    success,
                })
            }
            _ => Err(FunctionCallError::RespondToModel(
                "mcp handler received unexpected response variant".to_string(),
            )),
        }
    }
}

fn map_tool_error(err: ToolError) -> FunctionCallError {
    match err {
        ToolError::Rejected(message) => FunctionCallError::RespondToModel(message),
        ToolError::Codex(other) => {
            FunctionCallError::RespondToModel(format!("mcp tool call failed: {other:?}"))
        }
    }
}

File: src/nativeMain/kotlin/ai/solace/coder/core/tools/handlers/Mcp.kt
// port-lint: source core/src/tools/handlers/mcp.rs
package ai.solace.coder.core.tools.handlers

import ai.solace.coder.core.function_tool.FunctionCallError
import ai.solace.coder.core.tools.ToolHandler
import ai.solace.coder.core.tools.ToolInvocation
import ai.solace.coder.core.tools.ToolKind
import ai.solace.coder.core.tools.ToolOutput
import ai.solace.coder.core.tools.ToolPayload
import ai.solace.coder.protocol.CallToolResult
import ai.solace.coder.protocol.ResponseInputItem

class McpHandler : ToolHandler {
    override fun kind(): ToolKind {
        return ToolKind.Mcp
    }

    override suspend fun handle(invocation: ToolInvocation): ToolOutput {
        val payload = invocation.payload as? ToolPayload.Mcp ?: return ToolOutput.Mcp(
            ai.solace.coder.protocol.Result(
                value = null,
                error = "Invalid payload for McpHandler"
            )
        )

        val result = invocation.session.callMcpTool(
            payload.server,
            payload.tool,
            payload.rawArguments
        )

        return ToolOutput.Mcp(result)
    }
}
